pkg/provider/azure_vmss_cache.go

```diff
@@ -146,8 +146,16 @@ func extractVmssVMName(name string) (string, string, error) {
 func (ss *ScaleSet) getVMSSVMCache(resourceGroup, vmssName string) (string, *azcache.TimedCache, error) {
 	cacheKey := strings.ToLower(fmt.Sprintf("%s/%s", resourceGroup, vmssName))
 	if entry, ok := ss.vmssVMCache.Load(cacheKey); ok {
-		cache := entry.(*azcache.TimedCache)
-		return cacheKey, cache, nil
+		return cacheKey, entry.(*azcache.TimedCache), nil
+	}
+
+	// lock and try find cacheKey from cache again, refresh cache if still not found
+	lockKey := cacheKey + "/search"
+	ss.lockMap.LockEntry(lockKey)
```

Comment:
Have you considered using [`singleflight`](https://pkg.go.dev/golang.org/x/sync/singleflight) to suppress duplicate calls rather than adding a new lock?

Reply:
@edreed  using [singleflight](https://pkg.go.dev/golang.org/x/sync/singleflight) requires refactor of vmss cache, adding a new lock would be quite small and safe change, let's roll out this fix first

Reply:
It shouldn't require a refactor for use in this context, e.g.

```go
v, err, _ := ss.group.Do(cacheKey, func () (interface{}, error) {
	cache, err := ss.newVMSSVirtualMachinesCache(resourceGroup, vmssName, cacheKey)
	if err != nil {
		return nil, err
	}
	ss.vmssVMCache.Store(cacheKey, cache)
	return cache, nil
});
if err != nil {
	return "", nil, err
}
cache := v.(*azcache.TimedCache)
return cacheKey, cache, nil
```

where `ss.group` is a [`singleflight.Group`](https://pkg.go.dev/golang.org/x/sync/singleflight#Group).

Reply:
thanks, using [singleflight.Group](https://pkg.go.dev/golang.org/x/sync/singleflight#Group) now in the PR, I think this PR could fix the vmss list storm since it does not have lock in the before, lots of new vmss cache are created simultaneously which cause the vmss list storm.